File: LR1_Java/LR1.java
import java.util.*;

class LR1
{
    // Static tables for LR(1) parser
    static HashMap<Integer, HashMap<Character, String>> actionTable = new HashMap<>();
    static HashMap<Integer, HashMap<Character, Integer>> gotoTable = new HashMap<>();
    static HashMap<Integer, String> productions = new HashMap<>();
    static HashSet<Integer> states = new HashSet<>();

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        System.out.print("\nEnter number of productions: ");
        int p = sc.nextInt();
        sc.nextLine();

        System.out.println("\nEnter productions (Example: S=CC  OR  C=cC|d)");

        // Track production number separately to handle | expansions
        int prodIndex = 1;  

        for(int i = 1; i <= p; i++)
        {
            String prod = sc.nextLine();
            char lhs = prod.charAt(0);
            String rhs = prod.substring(prod.indexOf('=') + 1);

            // Split by '|' and store each alternative as a separate production
            String[] alternatives = rhs.split("\\|");

            for(String alt : alternatives)
            {
                productions.put(prodIndex, lhs + "=" + alt.trim());
                prodIndex++;
            }
        }

        System.out.println("\nEnter ACTION table entries");
        System.out.println("Format: state terminal action");
        System.out.println("Example: 0 c s3   OR   4 $ acc");
        System.out.println("Type 'end' to stop");

        while(true)
        {
            String line = sc.nextLine();

            if(line.equals("end"))
                break;

            String[] parts = line.split(" ");

            int state = Integer.parseInt(parts[0]);
            char terminal = parts[1].charAt(0);
            String action = parts[2];

            actionTable.putIfAbsent(state, new HashMap<>());
            actionTable.get(state).put(terminal, action);

            states.add(state);
        }

        System.out.println("\nEnter GOTO table entries");
        System.out.println("Format: state nonTerminal nextState");
        System.out.println("Example: 0 S 1");
        System.out.println("Type 'end' to stop");

        while(true)
        {
            String line = sc.nextLine();

            if(line.equals("end"))
                break;

            String[] parts = line.split(" ");

            int state = Integer.parseInt(parts[0]);
            char nt = parts[1].charAt(0);
            int next = Integer.parseInt(parts[2]);

            gotoTable.putIfAbsent(state, new HashMap<>());
            gotoTable.get(state).put(nt, next);

            states.add(state);
            states.add(next);
        }

        // Print expanded productions for user reference
        System.out.println("\n======================================");
        System.out.println("\tExpanded Productions");
        System.out.println("======================================");
        for(int k : new TreeSet<>(productions.keySet()))
            System.out.println("  " + k + ". " + productions.get(k));

        System.out.print("\nEnter input string: ");
        String input = sc.nextLine() + "$";

        printTables();
        parse(input);

        sc.close();
    }

    static void printTables()
    {
        TreeSet<Integer> sortedStates = new TreeSet<>(states);

        System.out.println("\n======================================");
        System.out.println("\tLR(1) ACTION TABLE");
        System.out.println("======================================");

        for(int s : sortedStates)
        {
            if(actionTable.containsKey(s))
            {
                for(char t : actionTable.get(s).keySet())
                {
                    System.out.println("ACTION[" + s + "," + t + "] = " +
                            actionTable.get(s).get(t));
                }
            }
        }

        System.out.println("\n======================================");
        System.out.println("\tLR(1) GOTO TABLE");
        System.out.println("======================================");

        for(int s : sortedStates)
        {
            if(gotoTable.containsKey(s))
            {
                for(char nt : gotoTable.get(s).keySet())
                {
                    System.out.println("GOTO[" + s + "," + nt + "] = " +
                            gotoTable.get(s).get(nt));
                }
            }
        }
    }

    static void parse(String input)
    {
        Stack<Integer> stateStack = new Stack<>();
        Stack<String> symbolStack = new Stack<>();

        stateStack.push(0);

        int i = 0;

        System.out.println("\n===========================================================================");
        System.out.printf("%-30s %-20s %-20s\n", "Stack", "Input", "Action");
        System.out.println("===========================================================================");

        while(true)
        {
            int state = stateStack.peek();
            char current = input.charAt(i);

            String action = null;

            if(actionTable.containsKey(state))
                action = actionTable.get(state).get(current);

            String stackStr = buildStackString(symbolStack, stateStack);
            String remainingInput = input.substring(i);

            if(action == null)
            {
                System.out.printf("%-30s %-20s ERROR\n", stackStr, remainingInput);
                System.out.println("\n===========================================================================");
                System.out.println("\t\tString Rejected");
                System.out.println("===========================================================================");
                return;
            }

            if(action.equals("acc"))
            {
                System.out.printf("%-30s %-20s ACCEPT\n", stackStr, remainingInput);
                System.out.println("\n===========================================================================");
                System.out.println("\t\tString Accepted");
                System.out.println("=============================================================================");
                return;
            }

            else if(action.startsWith("s"))
            {
                int nextState = Integer.parseInt(action.substring(1));

                System.out.printf("%-30s %-20s Shift %c\n", stackStr, remainingInput, current);

                symbolStack.push(String.valueOf(current));
                stateStack.push(nextState);

                i++;
            }

            else if(action.startsWith("r"))
            {
                int prodNum = Integer.parseInt(action.substring(1));
                String production = productions.get(prodNum);

                char lhs = production.charAt(0);
                String rhs = production.substring(production.indexOf('=') + 1);

                System.out.printf("%-30s %-20s Reduce by %s\n",stackStr, remainingInput, production);

                if(!rhs.equals("#"))
                {
                    for(int k = 0; k < rhs.length(); k++)
                    {
                        symbolStack.pop();
                        stateStack.pop();
                    }
                }

                symbolStack.push(String.valueOf(lhs));

                int topState = stateStack.peek();

                if(!gotoTable.containsKey(topState) || !gotoTable.get(topState).containsKey(lhs))
                {
                    System.out.printf("%-30s %-20s ERROR (missing GOTO[%d,%c])\n",
                            buildStackString(symbolStack, stateStack), remainingInput, topState, lhs);
                    System.out.println("\n===========================================================================");
                    System.out.println("\t\tString Rejected");
                    System.out.println("===========================================================================");
                    return;
                }

                int gotoState = gotoTable.get(topState).get(lhs);
                stateStack.push(gotoState);
            }
        }
    }

    static String buildStackString(Stack<String> symbolStack, Stack<Integer> stateStack)
    {
        StringBuilder sb = new StringBuilder();
        sb.append(stateStack.get(0));

        for(int k = 0; k < symbolStack.size(); k++)
        {
            sb.append(" ").append(symbolStack.get(k));
            sb.append(" ").append(stateStack.get(k + 1));
        }

        return sb.toString();
    }
}


/*

================ SAMPLE INPUT =================

Enter number of productions: 2

Enter productions:
S=CC
C=cC|d

ACTION TABLE
------------
0 c s3
0 d s4
1 $ acc
2 c s3
2 d s4
3 c s3
3 d s4
4 c r3
4 d r3
4 $ r3
5 $ r1
6 c r2
6 d r2
6 $ r2
end

GOTO TABLE
----------
0 S 1
0 C 2
2 C 5
3 C 6
end

Input: dd
Expected: String Accepted

================================================

*/

File: LR1/LR1.py
action_table = {}
goto_table = {}
productions = {}
states = set()


def build_stack_string(symbol_stack, state_stack):
    sb = str(state_stack[0])
    for k in range(len(symbol_stack)):
        sb += " " + symbol_stack[k] + " " + str(state_stack[k + 1])
    return sb


def print_tables():
    print("\n======================================")
    print("\tLR(1) ACTION TABLE")
    print("======================================")
    for s in sorted(states):
        for t in action_table.get(s, {}):
            print(f"ACTION[{s},{t}] = {action_table[s][t]}")

    print("\n======================================")
    print("\tLR(1) GOTO TABLE")
    print("======================================")
    for s in sorted(states):
        for nt in goto_table.get(s, {}):
            print(f"GOTO[{s},{nt}] = {goto_table[s][nt]}")


def parse(input_str):
    state_stack = [0]
    symbol_stack = []
    i = 0

    print("\n===========================================================================")
    print(f"{'Stack':<30} {'Input':<20} {'Action':<20}")
    print("===========================================================================")

    while True:
        state = state_stack[-1]
        current = input_str[i]
        action = action_table.get(state, {}).get(current)
        stack_str = build_stack_string(symbol_stack, state_stack)
        remaining = input_str[i:]

        if action is None:
            print(f"{stack_str:<30} {remaining:<20} ERROR")
            print("\n===========================================================================")
            print("\t\tString Rejected")
            print("===========================================================================")
            return

        if action == "acc":
            print(f"{stack_str:<30} {remaining:<20} ACCEPT")
            print("\n===========================================================================")
            print("\t\tString Accepted")
            print("=============================================================================")
            return

        elif action.startswith("s"):
            print(f"{stack_str:<30} {remaining:<20} Shift {current}")
            symbol_stack.append(current)
            state_stack.append(int(action[1:]))
            i += 1

        elif action.startswith("r"):
            prod_num = int(action[1:])
            production = productions[prod_num]
            lhs = production[0]
            rhs = production[production.index('=') + 1:]
            print(f"{stack_str:<30} {remaining:<20} Reduce by {production}")
            if rhs != "#":
                for _ in range(len(rhs)):
                    symbol_stack.pop()
                    state_stack.pop()
            symbol_stack.append(lhs)
            top_state = state_stack[-1]
            if top_state not in goto_table or lhs not in goto_table[top_state]:
                print(f"{build_stack_string(symbol_stack, state_stack):<30} {remaining:<20} ERROR")
                print("\n===========================================================================")
                print("\t\tString Rejected")
                print("===========================================================================")
                return
            state_stack.append(goto_table[top_state][lhs])


print("\nEnter number of productions: ", end="")
p = int(input())

print("\nEnter productions (Example: S=CC  OR  C=cC|d)")

prod_index = 1
for i in range(1, p + 1):
    prod = input()
    lhs = prod[0]
    rhs = prod[prod.index('=') + 1:]
    for alt in rhs.split("|"):
        productions[prod_index] = lhs + "=" + alt.strip()
        print(f"  Stored production {prod_index}: {lhs}={alt.strip()}")
        prod_index += 1

print("\nEnter ACTION table entries")
print("Format: state terminal action")
print("Example: 0 c s3   OR   4 $ acc")
print("Type 'end' to stop")

while True:
    line = input()
    if line == "end":
        break
    parts = line.split()
    state, terminal, action = int(parts[0]), parts[1][0], parts[2]
    action_table.setdefault(state, {})[terminal] = action
    states.add(state)

print("\nEnter GOTO table entries")
print("Format: state nonTerminal nextState")
print("Example: 0 S 1")
print("Type 'end' to stop")

while True:
    line = input()
    if line == "end":
        break
    parts = line.split()
    state, nt, next_state = int(parts[0]), parts[1][0], int(parts[2])
    goto_table.setdefault(state, {})[nt] = next_state
    states.update([state, next_state])

print("\n======================================")
print("\tExpanded Productions")
print("======================================")
for k in sorted(productions):
    print(f"  {k}. {productions[k]}")

print_tables()

print("\nEnter input string: ", end="")
parse(input() + "$")